### DISCUSSION TITLE
3rd Party API requires POST with "Content-Type: multipart/form-data" and an XML file

### DISCUSSION THREAD

**Discussion by cngi1** [Budibase/budibase#5511] [Help]
I am working with a 3rd party 3.0 API that is mostly JSON but to update some databases they have not completed PUT or POST with JSON in their API.  

They want a POST with "Content-Type: multipart/form-data" and an XML file.  
https://swagger.io/docs/specification/describing-request-body/multipart-requests/

The XML file begins with 
<?xml version="1.0" encoding="utf-8" ?>
<TestSpecification xmlns="TestSpecification.xsd">  // the XSD is on the server
Then the rest of the XML data they use for the update.

So in BB I would grab the original data, modify it and send it back with a POST and file.

I don't see how to create files for this requirement and I don't know enough about this process to fake the file and include contents in the POST?

Can this be done with BB?  

thank you in advance.

**Comment by melohagan** [comment]
Hi @cngi1 

You can post form-data:
![Screenshot 2022-04-21 at 15 29 33](https://user-images.githubusercontent.com/101575380/164480644-720a8310-0293-4a8d-be96-a9be3c4cde47.png)

And you can use bindings to fill out the key/value pairs: https://docs.budibase.com/docs/rest-bindings

However, I am not sure of a way to send files as part of the multipart/form-data within Budibase. I did find this related issue: https://github.com/Budibase/budibase/issues/1413

It was closed, but only by the stale bot, which no longer closes stale issues.

**Comment by cngi1** [comment]
How do you reopen or create a new issue?  Without the file sending capability my project is now DEAD.

**Comment by xnohat** [comment]
Workaround solution:

Create Automation:
Manipulate data as you need and pass it to a Bash Script block , Bash Script block will use CURL to upload file

Bash script of mine: this script will check curl command, install it, get data from previous step and pipe it to a file, then use curl to upload file to an api
```
command -v curl >/dev/null 2>&1 || { apt -y install curl; }
echo {{steps.2.value}} | base64 --decode > /app/elasticsearchbulkdata.json
curl -H 'Content-Type: application/x-ndjson' -XPOST 'http://elasticsearch:9200/shop/_bulk?pretty' --data-binary @/app/elasticsearchbulkdata.json
```

**Comment by cngi1** [comment]
thank you for the idea and example @xnohat,  Are you running BB on your server / VM? 

I haven't ventured of the Cloud BB.
